src/Map/MapRasterizer_P.cpp

```diff
@@ -488,17 +480,144 @@ void OsmAnd::MapRasterizer_P::rasterizePolyline(
                 if (prevCross != 0 || !intersect)
                 {
                     simplifyVertexToDirection(context, pVertex, vertex, tempVertex);
-                    path.moveTo(tempVertex.x, tempVertex.y);
+                    if (shift)
+                    {
+                        originalPoints.push_front(tempVertex);
+                        auto normal = Utilities::computeNormalToLine(pVertex, vertex, dir);
+                        tempVertex += normal * offset;
+                        shiftedPoints.push_front(tempVertex);
+                    }
+                    else
+                    {
+                        path.moveTo(tempVertex.x, tempVertex.y);
+                    }
                 }
                 simplifyVertexToDirection(context, vertex, pVertex, tempVertex);
-                path.lineTo(tempVertex.x, tempVertex.y);
+                if (shift)
+                {
+                    originalPoints.push_front(tempVertex);
+                    auto normal = Utilities::computeNormalToLine(pVertex, vertex, dir);
+                    tempVertex += normal * offset;
+                    shiftedPoints.push_front(tempVertex);
+
+                    // fix corner shifts
+                    if (originalPoints.size() >= kLastPointCnt)
+                    {
+                        // calculation of the angle between two outermost segments of the path.
+                        auto vecADir = (originalPoints[0] - originalPoints[1]).normalized();
+                        auto vecBDir = (originalPoints[2] - originalPoints[1]).normalized();
+                        auto angle = atan2(vecBDir.y, vecBDir.x) - atan2(vecADir.y, vecADir.x);
+                        if (angle > M_PI) {
+                            angle -= 2 * M_PI;
+                        } else if (angle <= -M_PI) {
+                            angle += 2 * M_PI;
+                        }
+                        // calculation of the angle between two outermost segments of the path.
```

Comment:
double same comment just cries that you need to create a method

Reply:
Moved to the dedicated function.